File: palindrome_linked_list.py
# 234. Palindrome Linked List

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if head == None:
        	return True
        else:
        	dummy = ListNode(None)
        	dummy.next = head
        	prev = dummy
        	#curr = head
        	slow = head
        	fast = head
        	while fast.next and fast.next.next:
        		fast = fast.next.next
        		tmp = slow.next
        		slow.next = prev
        		slow = slow.next
        		prev = slow
        		
        

File: palindrome_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode *midNode = getMid(head);
        ListNode *rhead = reverseList(midNode);
        while (rhead) {
            if (head->val != rhead->val) {
                return false;
            }
            head = head->next;
            rhead = rhead->next;
        }
        return true;
    }
    
    ListNode* reverseList(ListNode* head) {
        ListNode node(0);
        ListNode* L = &node;
        ListNode* q;
        while (head) {
            q = head->next;
            head->next = L->next;
            L->next = head;
            head = q;
        }
        return L->next;
    }
    
    ListNode* getMid(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        return slow;
    }
};